### PR TITLE
[FLINK-2191] Fix inconsistent use of closure cleaner in Scala Streaming

### PR DESCRIPTION
**Pull request by aljoscha**
The closure cleaner still cannot be disabled for the Timestamp extractor
in Time and for the delta function in Delta (windowing helpers).

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 1]
```diff
@@ -781,11 +780,24 @@ class DataStream[T](javaStream: JavaStream[T]) {
...
       def invoke(in: T) = cleanFun(in)
     }
     this.addSink(sinkFunction)
   }
 
+  /**
+   * Returns a "closure-cleaned" version of the given function. Cleans only if closure cleaning
+   * is not disabled in the {@link org.apache.flink.api.common.ExecutionConfig}
+   */
+  private[flink] def clean[F <: AnyRef](f: F): F = {
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala]: Why do we need multiple clean functions? I think scala visibility solves that. :)

File: flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala [hunk 2]
```diff
@@ -455,15 +455,22 @@ class StreamExecutionEnvironment(javaEnv: JavaEnv) {
...
+   * is not disabled in the {@link org.apache.flink.api.common.ExecutionConfig}
+   */
+  private[flink] def clean[F <: AnyRef](f: F): F = {
+    if (getConfig.isClosureCleanerEnabled) {
+      ClosureCleaner.clean(f, true)
+    } else {
+      ClosureCleaner.ensureSerializable(f)
+    }
+    f
+  }
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: I would suggest to have just this clean.

Comment (aljoscha) [inline flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala]: I will look into it. The problem is that the DataStream and all the other classes don't have a reference to the Scala StreamExecutionEnvironment.

**Comment by aljoscha** [comment]
I updated it, now we still have several "clean" functions but they just forward to the StreamingExEnv.scalaClean() function.

**Comment by gyfora** [comment]
Looks good to me :+1:

**Comment by mbalassi** [comment]
Merging.

**Pull request closed without merging**